### ISSUE TITLE
Option for force feedback center spring for wheel devices

### ISSUE THREAD

**Issue by supersmo** [x360ce/x360ce#123]
Enhancement request:
A lot of car games aren't designed for steering wheels and only support rumble effects (like F1 race stars) so when I hook up my steering wheels to these games with force feedback switched on, my wheels rumble but they loose the default center spring.

I'd love to have an option in x360ce to add center spring effect so that I can have both rumble and center spring in the games.

**Comment by EJocys** [comment]
Agree, I was thinking about implementing custom centring spring forces, too, because on Logitech G27 it is horrible. G27 have huge force dead-zone at the centre. I think, Logitech excuse that huge dead zone is required, because it prevents two motors fighting is BS. By monitoring current wheel position it is possible to get wheel movement speed and use that to counter oscillation/swinging problem and reduce centring spring force dead-zone close to zero.

**Comment by supersmo** [comment]
Nice to hear that you already thought of this. I was hoping for a center spring checkbox and a strength slider so even more customization options sounds fantastic. :-)

**Comment by Squall-Leonhart** [comment]
Only a single line of code is required to enable center spring, but customised spring would be a bit more.

**Comment by supersmo** [comment]
Personally I'd be thrilled for the one-liner. I think it's sufficient for my needs and it sounds like there's a bigger chance that it might be included in the next release :-)
Of course if there were a customized center spring I would play around with the settings.
Either way I would be very thankful. You are the programmers so I'll leave the implementation details to you :-)

**Comment by johan84** [comment]
Yes, In fact would be great if you could give centering spring to all axes of any joystick, currently I want to play Crysis, Ace Combat Assault Horizon, & other games which supports the gamepad Xbox 360 and the rumble vibration with my Ms Force Feedback 2 Joystick and when I try play Crysis, FFB effects are great when firing or lunch a granade but it's impossible to play because NO there are  self-centering effect, even it would be interesting to add that you could Customize the centering force with slider and future versions would also be great add Damping and Friction  as a software ForceTest.exe
![forcetest](https://cloud.githubusercontent.com/assets/12949944/8228694/763877e4-1576-11e5-80a7-18a88f64c301.png)
http://www.fs-force.com/support.htm. I hope these suggestions are implemented soon, meanwhile I will try put a physical spring adapted to self-centered my joystick, like other gamer do it: 
![full-27841-82015-20140620_135439](https://cloud.githubusercontent.com/assets/12949944/8228713/9a7f63a6-1576-11e5-9443-6b2829124fb9.jpg)
However I do not think it's easy to get the real spring and lines of code to achieve the effect on  xbox360ce can be much better.  ;)

**Comment by bservet** [comment]
I'd love to see this implemented as well. The older drivers of Fanatec had this feature but it has sadly been removed.

**Comment by Boomslangnz** [comment]
Will this be added at all? Im really keen to get center spring going on my wheel

**Comment by mickael28** [comment]
Hi guys, 

Any update on this? I've tried a few games and the wheel is quite dead without any kind of resistance in games which do not support it at present...

**Comment by DavDaddy** [comment]
I'd love to have this feature as well.  Any word on if it is in the pipeline?

**Comment by Kuiriel** [comment]
I have the same issue with my Microsoft Sidewinder ForceFeedback 2 (like the one in the picture above). Oddly enough it hasn't been a problem all the way. In GTA V I could merge my joystick and my ps4 controller and ended up having both rumble translated as forcefeedback while flying (which was so awesome to have back that I shed a tear) and return to center. I'm trying to get it working again for BF2042 so using the newest version and I've noticed two issues

If I start up X360CE (4.17.15.0 Build 2020-11-16) return to center is fine until I run the Force Feedback test. Usually both left and right motor tests will work with PeriodicSine, though sometimes adjusting Right Motor settings will make the window flash and no success in the test. Either way, after the test the joystick no longer returns to center (RTC). 

If I exit the program, the joystick will still not RTC. However it does RTC correctly if using and exiting Xoutput.

If I open X360CE and exit without doing force feedback tests, RTC continues to work normally. Feels weaker than after a normal PC reboot though.

So I re-open X360CE. Ran the force feedback tests and RTC worked this time. Dropped the test strength to zero, ran the tests again, joystick vibrated as if full strength, still RTC. Pushed test strength to 100, now right motor test does nothing, still RTC. Moved test strength to 50, now right motor test works but the joystick has lost all RTC. Exit program, all RTC remains lost.

I'm very curious as to what's falling through the cracks - but it seems to be happening specifically while running the tests.

Edit: But oh my god once again I've hit the holy grail, my force feedback joystick can actually have rumble converted into a force feedback effect which is so much fun for immersion. Thank you thank you thank you. It's all good as long as I don't mess with the Strength component before playing, looks like!

**Comment by Kranf99** [comment]
I am using x360ce to connect an old salvaged Fanatec wheel to the game "Need For Speed" for my 10 years old son. It's great fun. ..BUT the spring centering effect on the wheel is 100% missing and this is very disturbing when using the wheel.

The people that have a Logitech wheels are lucky: Their logitech driver allows them to directly add a spring centering effect to their wheel. ...but this option is not available for Fanatec wheels (and after talking with a few fanatec "fanboys" about the missing option, I think it will never be added).

So, could you just add a very simple setting, nothing fancy, just to center the wheel? Specifically, I need this option with the version 4.xx of x360ce because the older version does not work with the new Need for Speed games.

I think it will help *MANY* people! (i.e. all the people playing at the *many* Need for speed games and cyberpunk2077)

Many Thanks!

**Comment by neefflanders** [comment]
> I am using x360ce to connect an old salvaged Fanatec wheel to the game "Need For Speed" for my 10 years old son. It's great fun. ..BUT the spring centering effect on the wheel is 100% missing and this is very disturbing when using the wheel.
> 
> The people that have a Logitech wheels are lucky: Their logitech driver allows them to directly add a spring centering effect to their wheel. ...but this option is not available for Fanatec wheels (and after talking with a few fanatec "fanboys" about the missing option, I think it will never be added).
> 
> So, could you just add a very simple setting, nothing fancy, just to center the wheel? Specifically, I need this option with the version 4.xx of x360ce because the older version does not work with the new Need for Speed games.
> 
> I think it will help _MANY_ people! (i.e. all the people playing at the _many_ Need for speed games and cyberpunk2077)
> 
> Many Thanks!

this is my exact story, i got NFS ProStreet to work with 360ce all the FFB is there except for center spring. and its needed for higher end wheels.  all the detail is there except for the wheel wanting to re-center.. something that enforces center spring is desperately needed

**Comment by Gian9700** [comment]
> I am using x360ce to connect an old salvaged Fanatec wheel to the game "Need For Speed" for my 10 years old son. It's great fun. ..BUT the spring centering effect on the wheel is 100% missing and this is very disturbing when using the wheel.
> 
> The people that have a Logitech wheels are lucky: Their logitech driver allows them to directly add a spring centering effect to their wheel. ...but this option is not available for Fanatec wheels (and after talking with a few fanatec "fanboys" about the missing option, I think it will never be added).
> 
> So, could you just add a very simple setting, nothing fancy, just to center the wheel? Specifically, I need this option with the version 4.xx of x360ce because the older version does not work with the new Need for Speed games.
> 
> I think it will help _MANY_ people! (i.e. all the people playing at the _many_ Need for speed games and cyberpunk2077)
> 
> Many Thanks!

For spring force in games, you have the iracing ffb test, but when you combine that in x360 unfortunately the iracing app not working in game. GTA v enchanced online tested. Yo can choose for vibration and some ffb effects without spring force, or spring force without ffb effects and vibration sistem. I don't know if at the day can be possible activate spring since x360ce app.

All tested in CSL dd fanatec

**Comment by Gian9700** [comment]
> Only a single line of code is required to enable center spring, but customised spring would be a bit more.

Only development can make that? Only single line because the force and friction can be changed in fanatec app

**Comment by Kranf99** [comment]
..just a little message to say that after almost 1 year, I am still interested for the little patch to add "wheel centering" inside x360ce! Still waiting patiently!

**Comment by Gian9700** [comment]
last update in 2020 unfortly

**Comment by Gian9700** [comment]
> ..just a little message to say that after almost 1 year, I am still interested for the little patch to add "wheel centering" inside x360ce! Still waiting patiently!

In this link an user provides this link for download an alternative version from RACER_S creattor but i dont know if still compatible or outdated version 

https://forum.arcadecontrols.com/index.php?topic=152368.0

**Comment by EvoJocys** [comment]
Fixed. From the changelog:

4.21.30 — Added: Centering Spring holds a wheel at its centre. Off by default.
4.21.30 — Added: Auto finds the weakest Centering Spring that brings the wheel home.

Download: https://github.com/x360ce/x360ce/releases

Closing as part of a tracker clean-up after the 4.21 release. If it still happens on 4.21, reopen this issue with the version and what you see.
